### PR TITLE
Add To Cart with Options: Add the product selector in the Template sidebar

### LINKED ISSUE #53479: Add the Product Type Sidebar selector in the Template sidebar
It would be nice to show the product type at the Template Single Product level.
A proper place could be the document sidebar.
It will help show how the data propagation works and also serve as a UI/UX suggestion.
Later, we can define if it makes sense; otherwise, we should take a different path.

### PR DESCRIPTION
**Pull request by retrofox**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This adds a panel to the Single Product template document to switch the product type in a similar way to the Add To Cart with Options button does though its toolbar.

Closes #53479 #53207

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install and activate the WooCommerce Beta Tester plugin
2. Enable the `experimental-blocks` and `blockified-add-to-cart` features flags
  * Go to WooCommerce Admin Test Helper -> Features and enable `blockified-add-to-cart`

Screenshot 2024-12-05 at 8 18 00 AM
3. Edit the `Single Product` template
4. Confirm the document sidebar shows the `Product Type` panel, closed as default

Screenshot 2024-12-05 at 10 34 18 AM

5. Confirm you can open select another type by picking an option from the select component

image

6. You can add and focus the `Add To Cart with Options (Experimental)` block to show the toolbar
7. Confirm selected product type is consistent

https://github.com/user-attachments/assets/0c56abdb-fc03-416c-b893-b4cca89049ca

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by Aljullu — COMMENTED** [review]
Exciting to see this being worked on! Do you know why we need to register a plugin? Couldn't we simply use `<PluginDocumentSettingPanel>` directly from `edit.tsx`?

**Comment by retrofox** [comment]
> Do you know why we need to register a plugin? Couldn't we simply use `<PluginDocumentSettingPanel>` directly from `edit.tsx`?

Actually I haven't tried. The [doc is clear](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-document-setting-panel/) about how to extend the panel.

**Comment by juliaamosova** [comment]
@retrofox I believe Solaris was tagged as a reviewer here automatically due to original changes in `packages/js/e2e-utils-playwright/package.json` as seen [here](https://github.com/woocommerce/woocommerce/compare/8a1fb9bfaa471d8e31260b87f4862822c9561462..7601f5ee1b7e5333ae59baca69cee0139665d623) (removed space). If this is the only reason we've been tagged, we can be removed from being reviewers. Otherwise, please let us know if you are looking for a review from Solaris.

**Comment by retrofox** [comment]
> @retrofox I believe Solaris was tagged as a reviewer here automatically due to original changes in `packages/js/e2e-utils-playwright/package.json` as seen [here](https://github.com/woocommerce/woocommerce/compare/8a1fb9bfaa471d8e31260b87f4862822c9561462..7601f5ee1b7e5333ae59baca69cee0139665d623) (removed space). 

Removed the label @juliaamosova. Thanks, and sorry for the noise.

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/plugins/index.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import {
+	// @ts-expect-error no exported member.
+	PluginDocumentSettingPanel,
+} from '@wordpress/editor';
+/**
+ * Internal dependencies
+ */
+import useProductTypeSelector from '../hooks/use-product-type-selector';
+
+export const productTypeSelectorTitle = __( 'Product Type', 'woocommerce' );
```
Comment (octaedro) [inline plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/plugins/index.tsx]: Do we need to declare and export this `const`? I haven't seen it used anywhere else.

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/hooks/use-product-type-selector/index.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Hook to get data from the store to manage the product type selector.
+ * Also, it provides a function to switch the current product type.
+ * It's a layer of abstraction to the store.
+ *
+ * @return {ProductTypeSelector} The product type selector data and functions.
+ */
+export default function useProductTypeSelector(): ProductTypeSelector {
+	/*
+	 * Get the product types and the current product type
```
Comment (octaedro) [inline plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/hooks/use-product-type-selector/index.ts]: Nitpick: Can we modify this comment to be more specific? Maybe: `Get the available product types and the current product type`.

**Review by octaedro — APPROVED** [review]
Good job @retrofox! The code looks good and is testing well on my end!
I left a couple of small comments, but I'm going to pre-approve this PR.

**Review by Aljullu — COMMENTED** [review]
> Actually I haven't tried. The [doc is clear](https://developer.wordpress.org/block-editor/reference-guides/slotfills/plugin-document-setting-panel/) about how to extend the panel.

Fair enough, you're right, I didn't see the docs. Let's go with this approach, then. :slightly_smiling_face: 

I noticed a couple of things, though:

1. The product type selector is available in the template even if the new _Add to Cart with Options_ block isn't there. While not a strong opinion, I feel the selector should only be visible when a block using the store is in the template, otherwise it feels confusing. What do you think?



2. I'm seeing some new errors in the console after this PR. Can you reproduce as well?

**Comment by retrofox** [comment]
> Good job @retrofox! The code looks good and is testing well on my end! I left a couple of small comments, but I'm going to pre-approve this PR.

Thanks, Fernando, for the review. Going to address your feedback

**Comment by retrofox** [comment]
> 1. The product type selector is available in the template even if the new _Add to Cart with Options_ block isn't there. While not a strong opinion, I feel the selector should only be visible when a block using the store is in the template, otherwise it feels confusing. What do you think?

Agree. Let me take a look

**Comment by retrofox** [comment]
I've rebased the PR.

> 1. The product type selector is available in the template even if the new _Add to Cart with Options_ block isn't there.
> 2. I'm seeing some new errors in the console after this PR. Can you reproduce as well?

Both observations make sense, imo. I'd suggest addressing them in follow-ups. I could try to do it, but it may be a while.

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/plugins/index.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			onChange={ ( slug ) => set( slug ) }
+			help={ __(
+				'Switch product type to see how the template adapts to each one.',
+				'woocommerce'
+			) }
+		/>
+	);
+}
+
+export default function PluginDocumentSettingTemplateSelectorPanel() {
```
Comment (Aljullu) [inline plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/plugins/index.tsx]: What do you think about making this name a bit more specific, something like `ProductTypeSelectorPlugin` or something like that? With the current name, it's not clear to me it allows changing the product type.

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/index.tsx [hunk 4]
```diff
@@ -40,6 +42,14 @@ if ( shouldRegisterBlock ) {
 	// Select Simple product type
 	dispatch( woocommerceTemplateStateStore ).switchProductType( 'simple' );
 
+	// Extend editor, blocks, etc
```
Comment (Aljullu) [inline plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/index.tsx]: I guess this comment could be a bit more precise too? Something like:

```suggestion
	// Register a plugin that adds a product type selector to the template sidebar.
```

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/plugins/index.tsx (hunk 1 above)
Comment (Aljullu) [reply plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/plugins/index.tsx]: Yes, I think we can remove the export for now until we have a need to export it.

**Review by Aljullu — COMMENTED** [review]
Thanks for the updates, folks! :raised_hands: 

I can confirm this fixed the plugin warning, but I'm still seeing the one with the store: `Store "woocommerce/template-state" is already registered`. Can you reproduce too?

> Both observations make sense, imo. I'd suggest addressing them in follow-ups. I could try to do it, but it may be a while.

No worries, I think @octaedro and I can take care of it so you don't have to worry. :slightly_smiling_face:  @octaedro do you have any preference on who should take this PR?

I don't have a strong opinion about merging this PR as-is and working in follow-ups or fixing the comments before. I personally lean towards fixing them in this PR, this way we keep this feature in one single PR which will be easier to find in the future, but I'm fine either way.

File: plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/plugins/index.tsx (hunk 3 above)
Comment (octaedro) [reply plugins/woocommerce-blocks/assets/js/blocks/add-to-cart-with-options/plugins/index.tsx]: Good call! Done in the commit [a6f5fc6](https://github.com/woocommerce/woocommerce/pull/53480/commits/a6f5fc65ce9912589a718ac6be79b24a4f111ffd)

**Review by Aljullu — APPROVED** [review]
Thanks for the updates, @octaedro! I tested it and it works great. Code changes look good too.

**Pull request merged**